cdc/puller/file_sorter.go

```diff
@@ -0,0 +1,494 @@
+	x := old[n-1]
+	old[n-1] = nil
+	*h = old[0 : n-1]
+	return x
+}
+
+// readPolymorphicEvent reads a PolymorphicEvent from file reader and also advance reader
+// TODO: batch read
+func readPolymorphicEvent(rd *bufio.Reader) (*model.PolymorphicEvent, error) {
+	var byteLen [8]byte
+	n, err := rd.Read(byteLen[:])
+	if err != nil {
+		if err == io.EOF {
+			return nil, nil
+		}
+		return nil, errors.Trace(err)
+	}
+	if n < 8 {
+		return nil, errors.Errorf("invalid lenth data %s", byteLen)
+	}
+	dataLen := int(binary.BigEndian.Uint64(byteLen[:]))
+
+	data := make([]byte, dataLen)
+	n, err = io.ReadFull(rd, data)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != dataLen {
+		return nil, errors.Errorf("truncated data %s n: %d dataLen: %d", data, n, dataLen)
+	}
+
+	ev := &model.PolymorphicEvent{}
+	err = gob.NewDecoder(bytes.NewReader(data)).Decode(ev)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return ev, nil
+}
+
+func (fs *FileSorter) output(ctx context.Context, entry *model.PolymorphicEvent) {
+	select {
+	case <-ctx.Done():
+		return
+	case fs.outputCh <- entry:
+	}
+}
+
+func (fs *FileSorter) rotate(ctx context.Context, resolvedTs uint64) error {
+	// sortSingleFile reads an unsorted file into memory, sort in memory and rewritten
+	// sorted events ta a new file.
+	sortSingleFile := func(ctx context.Context, filename string) (string, error) {
+		fpath := filepath.Join(fs.dir, filename)
+		_, err := os.Stat(fpath)
+		if os.IsNotExist(err) {
+			return "", nil
+		}
+		data, err := ioutil.ReadFile(fpath)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		evs := make([]*model.PolymorphicEvent, 0)
+		idx := 0
+		for idx < len(data) {
+			dataLen := int(binary.BigEndian.Uint64(data[idx : idx+8]))
+			if idx+8+dataLen > len(data) {
+				return "", errors.New("unsorted file unexpected truncated")
+			}
+			ev := &model.PolymorphicEvent{}
+			err = gob.NewDecoder(bytes.NewReader(data[idx+8 : idx+8+dataLen])).Decode(ev)
+
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+			evs = append(evs, ev)
+			idx = idx + 8 + dataLen
+		}
+		sort.Slice(evs, func(i, j int) bool {
+			return evs[i].Ts < evs[j].Ts
+		})
+		batchFlushSize := 10
+		newfile := randomFileName("sorted")
+		newfpath := filepath.Join(fs.dir, newfile)
+		buffer := make([]*model.PolymorphicEvent, 0, batchFlushSize)
+		for _, entry := range evs {
+			buffer = append(buffer, entry)
+			if len(buffer) >= batchFlushSize {
+				_, err := flushEventsToFile(ctx, newfpath, buffer)
+				if err != nil {
+					return "", errors.Trace(err)
+				}
+				buffer = make([]*model.PolymorphicEvent, 0, batchFlushSize)
+			}
+		}
+		if len(buffer) > 0 {
+			_, err := flushEventsToFile(ctx, newfpath, buffer)
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+		}
+		return newfile, nil
+	}
+
+	// clear and reset unsorted files, set cache sorting flag to prevent repeated sort
+	fs.cache.fileLock.Lock()
+	if atomic.LoadInt32(&fs.cache.sorting) == 1 {
```

Comment:
it's needless that check the sorting is in process under lock protect
we can use `CompareAndSwap`  to avoid using lock